# doctests_lazy_matrix_mul.txt

>>> lazy_matrix_mul = __import__('101-lazy_matrix_mul').lazy_matrix_mul
>>> m_a = [[1, 2], [3, 4]]
>>> m_b = [[5, 6], [7, 8]]
>>> lazy_matrix_mul(m_a, m_b)
array([[19, 22],
       [43, 50]])

>>> lazy_matrix_mul()
Traceback (most recent call last):
...
TypeError: lazy_matrix_mul() missing 2 required positional arguments: 'm_a' and 'm_b'

>>> lazy_matrix_mul([[1, 2, 3], [4, 5, 6]])
Traceback (most recent call last):
...
TypeError: lazy_matrix_mul() missing 1 required positional argument: 'm_b'

>>> m_a = [[1, 2, 3], [4, 5, 6]]
>>> m_b = [[7], [8], [9]]
>>> lazy_matrix_mul(m_b, m_b)
Traceback (most recent call last):
...
ValueError: shapes (3,1) and (3,1) not aligned: 1 (dim 1) != 3 (dim 0)

>>> m_a = [[1.5, 2.8], [3.2, 4.7]]
>>> m_b = [[5.1, 6.3], [7.2, 8.4]]
>>> lazy_matrix_mul(m_a, m_b)
array([[27.81, 32.97],
       [50.16, 59.64]])

>>> m_a = [[1, 2], [3, 4]]
>>> m_b = [[5, 6, 7], [8, 9, 10]]
>>> lazy_matrix_mul(m_a, m_b)
array([[21, 24, 27],
       [47, 54, 61]])

>>> m_a = [[1, 2], [3, 4]]
>>> m_b = [[5, 6], [7, "8"]]
>>> lazy_matrix_mul(m_a, m_b)
Traceback (most recent call last):
    ...
TypeError: invalid data type for einsum

>>> lazy_matrix_mul([[1, 2], [3, 4]], [])
Traceback (most recent call last):
...
ValueError: shapes (2,2) and (0,) not aligned: 2 (dim 1) != 0 (dim 0)

>>> lazy_matrix_mul([], [[1, 2], [3, 4]])
Traceback (most recent call last):
...
ValueError: shapes (0,) and (2,2) not aligned: 0 (dim 0) != 2 (dim 0)

>>> lazy_matrix_mul([1, 2, 3], [[1, 2], [3, 4]])
Traceback (most recent call last):
...
ValueError: shapes (3,) and (2,2) not aligned: 3 (dim 0) != 2 (dim 0)

>>> lazy_matrix_mul([[1, 2], [3, 4]], [1, 2, 3])
Traceback (most recent call last):
...
ValueError: shapes (2,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)

>>> lazy_matrix_mul([[1, 2], [3, 4]], "string")
Traceback (most recent call last):
...
ValueError: Scalar operands are not allowed, use '*' instead

>>> lazy_matrix_mul("string", [[1, 2], [3, 4]])
Traceback (most recent call last):
...
ValueError: Scalar operands are not allowed, use '*' instead

>>> lazy_matrix_mul([[1, 2], [3, 4]], [[1, 2, 3], [4, 5]])
Traceback (most recent call last):
...
ValueError: setting an array element with a sequence.

>>> lazy_matrix_mul([[1, 2, 3], [4, 5]], [[1, 2], [3, 4]])
Traceback (most recent call last):
...
ValueError: setting an array element with a sequence.

>>> lazy_matrix_mul([[1, 2, 3], [4, 5, 6]], [[7, 8], [9, 10]])
Traceback (most recent call last):
...
ValueError: shapes (2,3) and (2,2) not aligned: 3 (dim 1) != 2 (dim 0)

